### PR TITLE
Support CUSTOM shader attributes in 2D

### LINKED ISSUE #6703: Allow passing custom vertex data to `canvas_item` shaders using a custom vertex array
### Describe the project you are working on

Complex 2D shaders that need a separate interpolation layer from UV (as UV is already in-use for texture mapping) for calculating implicit functions across the face of a triangle.

### Describe the problem or limitation you are having in your project

In Godot 3, I could use vertex colors in canvas item shaders as uncompressed float values that are interpolated across a triangle.

Now, in Godot 4, vertex colors are forced to be 8 bit integers, and there is no replacement for this kind of extra data for canvas item shaders. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Allow at least one of the 4 CUSTOM0-3 arrays to be accessed from canvas item shaders.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

https://docs.godotengine.org/en/4.0/tutorials/shaders/shader_reference/spatial_shader.html

https://docs.godotengine.org/en/4.0/tutorials/shaders/shader_reference/canvas_item_shader.html

Add CUSTOM0 to vertex shader of canvas item.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No workarounds.

### Is there a reason why this should be core and not an add-on in the asset library?

It cannot be done as an add-on.

### PR DESCRIPTION
**Pull request by Giwayume**
History: https://github.com/godotengine/godot/pull/76276

Closes: https://github.com/godotengine/godot-proposals/issues/6703

A general summary of changes:
- The vertex shader attributes CUSTOM0, CUSTOM1 are now supported for canvas item shaders for all renderers
- The vertex shader attributes CUSTOM2, CUSTOM3 are not supported in canvas item shaders in order to save attribute slots for future use

CUSTOM0 and CUSTOM1 vertex attributes are placed in the same attribute slot as their 3D counterparts. This simplifies the changes, so we don't have to modify things like the bit shift mask variables, or remap them.

In OpenGL ES 3, attribute layout slots 14 & 15 should exist under the minimum GL specification, so all of the slots are shifted down to make room for slots 6 & 7 to be used for custom attributes. You can read more about this in the proposal linked above.

Sample project:
[CanvasShaderTest.zip](https://github.com/godotengine/godot/files/13835818/CanvasShaderTest.zip)

### CODE REVIEW

**Comment by Giwayume** [comment]
@Calinou Please elaborate what you're looking for in a testing project beyond the sample project attached.

**Comment by AThousandShips** [comment]
Your branch is still garbled, you still have a merge commit as your commit, this isn't correct, please make a new commit with these changes and remove the old one 🙂

**Comment by Giwayume** [comment]
How are you checking this? The history looks clean to me. https://github.com/Giwayume/godot/commits/feature/canvas-item-shader-custom-data/

**Comment by AThousandShips** [comment]
You are one commit ahead and one behind master, this is wrong, and your commit has two parents, also incorrect, see https://github.com/Giwayume/godot/commit/507e2b0c29ef6235ff59ef630b70eedf7cc34905

**Comment by AThousandShips** [comment]
Now solved 🙂

**Comment by Giwayume** [comment]
I added the CUSTOM0-3 properties to 2D visual shader. Missed that initially.

The code is still working for forward/mobile renderer, but stopped working for compatibility after the merge.

Looks like something changed where the input_mask passed into mesh_storage->mesh_instance_surface_get_vertex_arrays_and_format is always static for 2D meshes.

So, I first dived into trying to add `vertex_input_mask` to CanvasShaderData, just like how SceneShaderData works, manipulating `CanvasShaderData::set_code` and all, but I've come to the conclusion that it may not be worth it. All the extra code to read the materials just to ultimately prevent setting default values for a couple of vertex attributes is probably not worth the cost. What do you think?

I just added the customs to the end of the input mask for every 2D mesh in GLES3:
`uint64_t input_mask = RS::ARRAY_FORMAT_VERTEX | RS::ARRAY_FORMAT_COLOR | RS::ARRAY_FORMAT_TEX_UV | RS::ARRAY_FORMAT_CUSTOM0 | RS::ARRAY_FORMAT_CUSTOM1;`

**Comment by Giwayume** [comment]
If this PR hangs for another couple of months with no input like the last one, I'm giving up.

**Comment by fire** [comment]
I called a few people to review.

**Comment by clayjohn** [comment]
This is worth some discussion with others in the rendering team as I am unsure what to do with this. 

An important design goal for us is to make basic rendering features compatible across all backends. Users should be able to rely on basic features continuing to work even when switching to the compatibility renderer. We try to limit the number of things that aren't supported in all backends. For the most part, things that aren't supported by all backends are optional features that can be turned on/off without significantly breaking the visuals (i.e. SDFGI, volumetric fog, etc.). 

This feature is a bit different. If we move forward with it, users will be able to create assets that work only in the RD-based backends. If they convert their project to the compatibility backend, their meshes will stop working entirely and they will have to re-design their assets to continue working. 

I don't think that is a situation I am comfortable introducing into the engine. Accordingly, in my mind, the only acceptable options are:
1. Only support CUSTOM0 and CUSTOM1 in the canvas item shader
2. Only support CUSTOM0 in the canvas item shader
3. Don't support any CUSTOM attributes in the canvas item shader

Seeing as the original proposal is only asking for one custom channel. It may make sense to just expose one custom value for now, and then expose another if there is demand. The trouble with exposing as many as possible is that we can't go backwards (i.e. we can always add more custom attributes, but we can never remove custom attributes).

This should be discussed with other rendering contributors in our rendering meetings

**Comment by Giwayume** [comment]
>     1. Only support CUSTOM0 and CUSTOM1 in the canvas item shader
> 
>     2. Only support CUSTOM0 in the canvas item shader
> 
>     3. Don't support any CUSTOM attributes in the canvas item shader

What would be the reason to pick 2 or 3 over 1? Saving vertex attribute slots for future unknown use?

**Comment by clayjohn** [comment]
> > ```
> > 1. Only support CUSTOM0 and CUSTOM1 in the canvas item shader
> > 
> > 2. Only support CUSTOM0 in the canvas item shader
> > 
> > 3. Don't support any CUSTOM attributes in the canvas item shader
> > ```
> 
> What would be the reason to pick 2 or 3 over 1? Saving vertex attribute slots for future unknown use?

That's right. If users only need one CUSTOM slot, then there isn't much point in exposing 2, especially if it might block us from adding a different feature down the road

**Comment by QbieShay** [comment]
I read this when fire pinged us and I was unsure myself. How does this interact with potentially exposing instance customs for canvas items? Are they sharing this resource?

Regarding the compatibly between Rd and compat renderer, I do think that having assets that work in Rd and not in compat is a big problem, especially since our default renderer is Rd and this is about 2D. I think users will stumble on this way too much: they decide to make a 2D game that has a couple of effects, start in Rd, then decide to publish on low end, switch to compat, everything breaks. For me 2d is more critical than 3D for compatibility.

**Comment by Giwayume** [comment]
> How does this interact with potentially exposing instance customs for canvas items? Are they sharing this resource?

In the case of Mesh, location 5 (instance_color_custom_data) is unused and in the case of both Mesh and MultiMesh, location 6 CUSTOM0 could be used.

**Comment by fire** [comment]
Anther possibility to do the the complexity tasks behind reorganizing these attributes in the code so that we get all CUSTOM variables for compatibility renderer, but it would good to estimate and get support from the rendering team about merging before doing the work.

**Comment by Giwayume** [comment]
> Another possibility to do the the complex tasks behind reorganizing these attributes in the code so that we get all CUSTOM variables for compatibility renderer and the forward renderer, but it would good to estimate and get support from the rendering team about merging before doing the work.

GLES3 is nearly at the limit for canvas item attributes even with a reorg, if there is a concern about blocking new functionality that would still apply. If all CUSTOM0-3 were supported, that leaves 1 attribute slot left over.

**Comment by clayjohn** [comment]
We discussed this yesterday in our weekly rendering meeting. We agreed that exposing CUSTOM0 and CUSTOM1 would be a good first step. In looking at the code closer we realized that we can free up one more slot if we need to by combining VERTEX and UV into one attribute. Therefore, we can safely expose CUSTOM0 and CUSTOM1 while having room for the future (which I think we will need quite soon actually). 

In theory we can expose more slots if we create a fallback shader that doesn't use attributes for batching properties. However, doing so will be a good amount of work and will make the code much more complex. So we prefer to only do that if we absolutely have to. Since the current users aren't asking for more than 1 CUSTOM vector, exposing 2 should be more than enough. 

In summary, for this PR to be merged, it just needs to be updated to only expose CUSTOM0 and CUSTOM1 for both RD renderers and the compatibility renderer.

**Comment by Giwayume** [comment]
I removed CUSTOM2 and CUSTOM3, updated demo project with MultiMesh sample.

[CanvasShaderTest.zip](https://github.com/godotengine/godot/files/13835818/CanvasShaderTest.zip)

**Comment by lyuma** [comment]
Just wanted to leave a note here as well in case people do need more than 8 channels of custom data in 3D...

Because CUSTOM0-3 can support 4-component full 32-bit floats, there is nothing stopping a developer from bitpacking attribute data from individual 16-bit half floats or even 8-bit integers (which is sufficient for colors), in which case you could possibly support 24-32 custom 8-bit channels (depending on how `floatBitsToUint` is used in your shader and how careful you are to avoid NaN)

I can't find any examples of people using this technique in practice, I suspect because people usually don't reach attribute limits... but anyway just want to put it out there. (and as always, be careful about perforamnce of vertex operations on old mobile hardware, especially with texture coords)

**Comment by Giwayume** [comment]
Is anyone scheduled to review this?

**Comment by clayjohn** [comment]
> Is anyone scheduled to review this?

Not scheduled, no. But I started reviewing it and got pulled away to other tasks. I aim to finish the review shortly

File: drivers/gles3/rasterizer_canvas_gles3.cpp [hunk 1]
```diff
@@ -1412,11 +1414,27 @@ void RasterizerCanvasGLES3::_render_batch(Light *p_lights, uint32_t p_index, Ren
...
+					material_data = static_cast<GLES3::CanvasMaterialData *>(material_storage->material_get_data(material, RS::SHADER_CANVAS_ITEM));
+				}
+				if (material_data) {
+					if (material_data->shader_data->version.is_valid() && material_data->shader_data->valid) {
+						shader_data = material_data->shader_data;
+						if (shader_data) {
+							vertex_input_mask = shader_data->vertex_input_mask;
+						}
+					}
+				}
```
Comment (clayjohn) [inline drivers/gles3/rasterizer_canvas_gles3.cpp:1432]: A couple of things here:
1. `material_data` is already available through the batch, It can be accessed with `state.canvas_instance_batches[p_index].material_data` so we can skip the first `if` statement
2. `vertex_input_mask` comes from the material. So instead of doing all these checks in the hot loop, we should do them when the batch is first created. Accordingly, I would add `vertex_input_mask` into the batch data structure and then cache it when caching `material` and `material_data`:

https://github.com/godotengine/godot/blob/6fea273ed3df7d4be9674d35aae698731fa823ea/drivers/gles3/rasterizer_canvas_gles3.cpp#L628-L629

File: drivers/gles3/storage/material_storage.cpp [hunk 2]
```diff
@@ -2601,6 +2614,12 @@ void CanvasShaderData::set_code(const String &p_code) {
 	MaterialStorage::get_singleton()->shaders.canvas_shader.version_set_code(version, gen_code.code, gen_code.uniforms, gen_code.stage_globals[ShaderCompiler::STAGE_VERTEX], gen_code.stage_globals[ShaderCompiler::STAGE_FRAGMENT], gen_code.defines, texture_uniform_data);
 	ERR_FAIL_COND(!MaterialStorage::get_singleton()->shaders.canvas_shader.version_is_valid(version));
 
+	vertex_input_mask = RS::ARRAY_FORMAT_VERTEX;
+	vertex_input_mask |= uses_color << RS::ARRAY_COLOR;
+	vertex_input_mask |= uses_uv << RS::ARRAY_TEX_UV;
```
Comment (clayjohn) [inline drivers/gles3/storage/material_storage.cpp:2619]: Color and UV are always used by the shader, so they should always be present in the mask. Godot uses them internally, even if users don't access them in their shader. So they have to be present in the mask. 

You can remove the `usage_flag_pointers` for them above too.

**Review by clayjohn — COMMENTED** [review]
Looks great so far!

I tested locally and can confirm that it works. I left some comments on cleaning up the code a bit and optimizing out some unnecessary work. Please let me know if you have any questions or if any of my suggestions don't make sense.

File: servers/rendering/shader_types.cpp [hunk 3]
```diff
@@ -250,6 +250,8 @@ ShaderTypes::ShaderTypes() {
 	shader_modes[RS::SHADER_CANVAS_ITEM].functions["vertex"].built_ins["VERTEX_ID"] = constt(ShaderLanguage::TYPE_INT);
 	shader_modes[RS::SHADER_CANVAS_ITEM].functions["vertex"].built_ins["AT_LIGHT_PASS"] = constt(ShaderLanguage::TYPE_BOOL);
 	shader_modes[RS::SHADER_CANVAS_ITEM].functions["vertex"].built_ins["TEXTURE_PIXEL_SIZE"] = constt(ShaderLanguage::TYPE_VEC2);
+	shader_modes[RS::SHADER_CANVAS_ITEM].functions["vertex"].built_ins["CUSTOM0"] = ShaderLanguage::TYPE_VEC4;
+	shader_modes[RS::SHADER_CANVAS_ITEM].functions["vertex"].built_ins["CUSTOM1"] = ShaderLanguage::TYPE_VEC4;
```
[the remark this answers is not in the document]
Comment (Giwayume) [reply servers/rendering/shader_types.cpp:254]: Is this wrong on the Spatial side too? That is what I based it from.

**Comment by Giwayume** [comment]
@clayjohn your feedback implemented. I believe I cleaned up all the unnecessary variables now.

File: servers/rendering/shader_types.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (clayjohn) [reply servers/rendering/shader_types.cpp]: In my opinion, yes its wrong. However, to make the variables constant would technically break compatibility as users could do the following:
```
void vertex() {
	CUSTOM0 = some_vec4;
	...
	some_other_variable = CUSTOM0;
}
```
Its kind of a crappy situation as we either break compatibility or we have inconsistent behavior between the shaders

**Review by clayjohn — APPROVED** [review]
Looks great to me! Good work! 

Thanks for sticking through multiple discussions and suggested changes. I think we settled on a good compromise that should make everyone happy.
